File: tests/unit-tests/file_tests/non-deterministic/clock_gettime_highlevel.c
#include <stdio.h>
#include <time.h>

/*
    This test indirectly calls the clock_gettime syscall through the clock() function.
    The clock() function in glibc internally uses clock_gettime to measure CPU time 
    consumed by the process.

    The program measures the CPU time taken to run 1,000,000 iterations of a simple loop.
    It records the start time using clock(), performs the computation, then records the
    end time and calculates the elapsed CPU time.
*/

int main() {
    // Get the start time using clock()
    clock_t begin = clock();

    printf("Running 1,000,000 iterations...\n");

    volatile long long sum = 0;
    for (long long i = 0; i < 1000000; i++) {
        sum += i;
    }

    // Get the end time using clock()
    clock_t end = clock();

    // Calculate elapsed time in seconds
    double elapsed_time = (double)(end - begin) / CLOCKS_PER_SEC;

    // Display results
    printf("\nStart time: %lld clock ticks\n", begin);
    printf("End time: %lld clock ticks\n", end);
    printf("Elapsed CPU time: %.9f seconds\n", elapsed_time);

    return 0;
}

File: tests/unit-tests/file_tests_rust/non-deterministic/clock_gettime_highlevel.rs
use std::time::{Instant, SystemTime, UNIX_EPOCH};

fn main() {
    // Get the start time using Instant
    let begin = Instant::now();
    let start_system_time = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();

    println!("Running 1,000,000 iterations...");

    let mut sum: i64 = 0;
    for i in 0..1_000_000 {
        sum += i;
    }

    // Get the end time
    let end = Instant::now();
    let end_system_time = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();

    // Calculate elapsed time
    let elapsed_time = end.duration_since(begin);

    // Display results
    println!("\nStart time: {} clock ticks", start_system_time);
    println!("End time: {} clock ticks", end_system_time);
    println!("Elapsed CPU time: {:.9} seconds", elapsed_time.as_secs_f64());
} 